### ISSUE TITLE
Multi-tenant / Cross-tenant authorization not working

### ISSUE THREAD

**Issue by bensussman** [Azure/go-autorest#583]
I am attempting to get "Cross Tenant Image Sharing" to work by following this guide: https://docs.microsoft.com/en-us/azure/virtual-machines/linux/share-images-across-tenants I am able to follow the guide verbatim, and spin up a VM successfully in "Tenant 2" using the CLI as described in the final "Create a VM using Azure CLI" step in the documentation.

When I attempt to create a VM using the golang CLI, the only way that I see to be "authorized" as a service principal with TWO separate tenants simultaneously is via the `AuxTenants` field on the `ClientCredentialsConfig` like so:

```
clientCreds := auth.NewClientCredentialsConfig(
	sp.ClientID,
	sp.ClientSecret,
	tenantTwoID,
)
clientCreds.AuxTenants = append(clientCreds.AuxTenants, tenantOneID)
multiTenantAuthorizer, e := clientCreds.Authorizer()
...
multiTenantVMClient := compute.NewVirtualMachinesClient(tenantTwoSubscriptionID)
multiTenantVMClient.Authorizer = multiTenantAuthorizer
multiTenantVMClient.Sender = httpClient
...
vmFuture, err := multiTenantVMClient.CreateOrUpdate(ctx, ....)
```

The behavior I see is:

1. `multiTenantVMClient.CreateOrUpdate` hangs for a very very very long time. (about TEN MINUTES)
2. `multiTenantVMClient.CreateOrUpdate` after hanging errors with:

```
Azure CreateOrUpdate() error: azure.multiTenantSPTAuthorizer#WithAuthorization: Failed to refresh one or more Tokens for request to https://management.azure.com/subscriptions/<tenant 2 subscription>/resourceGroups/<rg-name>/providers/Microsoft.Compute/virtualMachines/vm-12345?api-version=2019-03-01: StatusCode=0 -- Original Error: failed to refresh auxiliary token: adal: Refresh request failed. Status Code = '401'. Response body: {"error":"invalid_client","error_description":"AADSTS7000215: Invalid client secret is provided.\r\nTrace ID: 2f...f00
Correlation ID: 5e....b
Timestamp: 2020-11-25 19:03:50Z","error_codes":[7000215],"timestamp":"2020-11-25 19:03:50Z","trace_id":"2f9...0","correlation_id":"5e...b","error_uri":"https://login.microsoftonline.com/error?code=7000215"}
```

Note that this client secret is NOT invalid, if i remove the `AuxTenant` and update the image to be a public one, it works with the exact same Service Principal IDs and Secret. 

This appears to be a bug with go-autorest support of multi-tenant SPs

**Comment by bensussman** [comment]
@jhendrixMSFT Could you take a look at this please?

**Comment by jhendrixMSFT** [comment]
I'm taking a look.

FYI Terraform uses this feature (it was originally implemented for them) so it works in at least some scenarios.

**Comment by jhendrixMSFT** [comment]
The hang is due to improper error chain walking, leading to retrying the authorization request which will always fail.  I have a fix for this.

**Comment by bensussman** [comment]
Thanks @jhendrixMSFT for the quick fix.

As I am banging my head against this to get it to *actually work*, I am able to reproduce the "Invalid Secret" error in the CLI when doing 
```
az login --service-principal -u '<app ID>' -p '<Secret>' --tenant '<tenant 2 ID>'
```
As described in https://docs.microsoft.com/en-us/azure/virtual-machines/linux/share-images-across-tenants

So my conclusion is that this go package is working as intended (aside from mis-reporting the error), and the issue is much more generally with AAD cross-tenant-service principals.

My organization ( https://spell.ml ) is hoping to bring Azure support up to parity with AWS+GCP, and the way our system is constructed in those clouds is very straightforward: Public images, a custom AWS-IAM-Role / GCP-Service-Account endowed with sufficient privileges to create and manage VMs, and in AWS we use STS impersonation to execute operations as the IAM-Role, while in GCP we have direct acess to the Service Account. We are attempting to design our Azure permissions system in a similar way, but this "Cross Tenant VM Image Sharing" issue is causing major problems for us. 

The documentation for AAD cross-tenant login is framed around solving the problem of many-users logging in to an "application" that uses AAD as a SSO (see https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-convert-app-to-be-multi-tenant ). However to get "Cross Tenant VM Image Sharing" to work, we have to use some of these verbs (like oauth2/authorize for example) in a context where this "Application" is not *actually* "delegating" any permissions, but needs to have these "delegated API permissions: user.read"  in order be be authenticated in order to unlock this "multi-tenant-service-principal-login" feature that is a requirement for "Cross Tenant VM Image Sharing".

Ultimately these feels like a extra-confusing design choice that Azure made, but i was wondering if you knew of some documentation or explanation or guide that explained in sufficient detail what exactly all these AAD steps are doing, and why they are required, and how to set up a system like ours @jhendrixMSFT ? ( I have tried opening support tickets with Azure and Azure-for-Startups, and only spent tens of hours on the phone with help-desk contractors who don't understand AAD).

**Comment by jhendrixMSFT** [comment]
@bensussman I'm asking around internally to see who can help with this.

**Comment by jhendrixMSFT** [comment]
@bensussman was this ever resolved?

**Comment by bensussman** [comment]
Yes! It was resolved by switching the authorization details from the ServicePrincipal to the Application (or vice versa, honestly it's all very confusing). We did some phone calls directly with Microsoft Support around the onerous end-to-end workflow and got some great news that by this Summer a feature will be released allowing Azure customers to share VM Images publicly, or directly share them with particular subscriptions without going through the Cross-Tenant-Authorization flow (which even after this particular issue was solved, still requires the manual "authorization" step and so can't be fully automated). We are excited for this release, as it will simplify our workflow and allow us to remove this code.
